tutorials/sphinx_tuto/tensorclass_fashion.py

```diff
@@ -0,0 +1,272 @@
+# In this tutorial we demonstrate how tensorclasses can be used to
+# efficiently and transparently load and manage data inside a training
+# pipeline. The tutorial is based heavily on the `PyTorch Quickstart
+# Tutorial <https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html>`__,
+# but modified to demonstrate use of tensorclass. See the related tutorial using
+# ``TensorDict``.
+
+
+import torch
+import torch.nn as nn
+from tensordict import MemmapTensor
+from tensordict.prototype import tensorclass
+from torch.utils.data import DataLoader
+from torchvision import datasets
+from torchvision.transforms import ToTensor
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+print(f"Using device: {device}")
+
+
+###############################################################################
+# The ``torchvision.datasets`` module contains a number of convenient pre-prepared
+# datasets. In this tutorial we'll use the relatively simple FashionMNIST dataset. Each
+# image is an item of clothing, the objective is to classify the type of clothing in
+# the image (e.g. "Bag", "Sneaker" etc.).
+
+training_data = datasets.FashionMNIST(
+    root="data",
+    train=True,
+    download=True,
+    transform=ToTensor(),
+)
+test_data = datasets.FashionMNIST(
+    root="data",
+    train=False,
+    download=True,
+    transform=ToTensor(),
+)
+
+###############################################################################
+# Tensorclasses are dataclasses that expose dedicated tensor methods over
+# its contents much like ``TensorDict``. They are a good choice when the
+# structure of the data you want to store is fixed and predictable.
+#
+# As well as specifying the contents, we can also encapsulate related
+# logic as custom methods when defining the class. In this case we’ll
+# write a ``from_dataset`` classmethod that takes a dataset as input and
+# creates a tensorclass containing the data from the dataset. We create
+# memory-mapped tensors to hold the data. This will allow us to
+# efficiently load batches of transformed data from disk rather than
+# repeatedly load and transform individual images.
+
+
+@tensorclass
+class FashionMNISTData:
+    images: torch.Tensor
+    targets: torch.Tensor
+
+    @classmethod
+    def from_dataset(cls, dataset, device=None):
+        data = cls(
+            images=MemmapTensor(
+                len(dataset), *dataset[0][0].squeeze().shape, dtype=torch.float32
+            ),
+            targets=MemmapTensor(len(dataset), dtype=torch.int64),
+            batch_size=[len(dataset)],
+            device=device,
+        )
+        for i, (image, target) in enumerate(dataset):
+            data[i] = cls(images=image, targets=torch.tensor(target), batch_size=[])
+        return data
+
+    def __len__(self):
+        # need to define explicitly for `len()` to work
+        return self.batch_size[0] if self.batch_dims else 0
```

Comment:
This is a bit of a patch that ideally would be fixed properly inside the `tensorclass` decorator.